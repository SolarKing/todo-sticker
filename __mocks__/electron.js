// @flow
let listeners: Map<string, Function> = new Map();

class IpcMain {   

    on(channel: string, listener: Function) {
        listeners.set(channel, listener);
    }
    
}

class IpcRenderer {   

    send(channel: string, ...args: Array<mixed>) {
        const listener = listeners.get(channel);
        if (listener != null) {
            listener({}, ...args);
        }
    }
    
}

class BrowserWindow {
    _listeners: Map<string, Function> = new Map();

    webContents = {
        openDevTools() {

        },
        on(channel: string, listener: Function) {
            listeners.set(channel, listener);
        },
        emit(channel: string, ...args: Array<mixed>) {
            const listener = listeners.get(channel);
            if (listener != null) {
                listener(...args);
            }
        },
        destroy() {

        }
    }

    constructor() {
        this._listeners = new Map();
    }
    isMaximized() {
        return false;
    }
    loadURL() { }
    close() { }
    maximize() { }
    unmaximize() { }
    minimize() { }

    on(channel: string, listener: Function) {
        this._listeners.set(channel, listener);
    }

    emit(channel: string, ...args: Array<mixed>) {
        const listener = this._listeners.get(channel);
        if (listener != null) {
            listener(...args);
        }
    }

    static getFocusedWindow(): BrowserWindow {
        return new BrowserWindow();
    }

}

const ipcMain = new IpcMain();
const ipcRenderer = new IpcRenderer();

export {
    ipcMain,
    ipcRenderer,
    BrowserWindow
};