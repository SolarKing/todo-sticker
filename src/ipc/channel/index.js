// @flow
export const IpcChannels = {
    noteDescriptionSend: 'noteDescriptionSend',
};