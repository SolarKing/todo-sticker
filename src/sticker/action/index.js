// @flow

export class TodoDescriptionChangedIPC {
    id: string;
    description: string;
}

export type CloseTodoNoteIpcAction = {
    id: string;
}