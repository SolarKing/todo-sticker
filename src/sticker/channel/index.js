// @flow
export const IpcChannels = {
    noteDescriptionSend: 'noteDescriptionSend',
    todoDescriptionChanged: 'todoDescriptionChanged',
    closeTodoNote: 'closeTodoNote'
};