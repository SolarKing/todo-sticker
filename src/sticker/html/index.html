<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="UTF-8">
	<link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.7.2/css/all.css" integrity="sha384-fnmOCqbTlWIlj8LyTjo7mOUStjsKC4pOpQbqyi7RrhN7udi9RwhKkMHpvLbHG9Sr" crossorigin="anonymous">
	<title>React Setup</title>
</head>
<body>
	<!-- 欲插入 React Component 的位置 -->
    <div id="app"></div>
    <!-- <script type="text/javascript" src="index_bundle.js"></script> -->
</body>
</html>
