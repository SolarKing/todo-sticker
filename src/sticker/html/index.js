// @flow
import React from 'react';
import ReactDOM from 'react-dom';
import StickerDescription from '../../component/stickerDescription';
import { injectGlobal } from 'styled-components';

const app = document.getElementById('app');

app && ReactDOM.render(
    <StickerDescription />,
    app
);

injectGlobal`
    body {
        width: 100vw;
        height: 100vh;
        margin: 0;
        background-color: #f7f7a3;
        overflow-wrap: break-word;
        overflow-x: hidden;
        -webkit-user-select: none;
        -webkit-app-region: drag;
    }
`;


