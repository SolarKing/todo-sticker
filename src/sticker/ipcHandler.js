// @flow
import { ipcMain, BrowserWindow, Electron } from 'electron';
import { IpcChannels } from './channel';
import type { CreateStickerAction, EditStickerAction, DeleteStickerAction } from './action';

declare var __dirname: any;
declare var process: any;

class StickerHandler {

    notes: Map<string, BrowserWindow>;
    _createNoteAndSendDescription: Function;
    _updateTodoNoteDescription: Function;
    _closeTodoNote: Function;
    _windowOptions: any = {
        width: 350, 
        height: 300, 
        frame: false, 
        titleBarStyle: 'hidden',
        minWidth: 180,
        minHeight: 150,
    }

    constructor() {
        this.notes = new Map();
        this._createNoteAndSendDescription = this._createNoteAndSendDescription.bind(this);
        this._updateTodoNoteDescription = this._updateTodoNoteDescription.bind(this);
        this._closeTodoNote = this._closeTodoNote.bind(this);
    }

    registerListener() {
        ipcMain.on(IpcChannels.createSticker, this._createNoteAndSendDescription);
        ipcMain.on(IpcChannels.editSticker, this._updateTodoNoteDescription);
        ipcMain.on(IpcChannels.deleteSticker, this._closeTodoNote);
    }

    _createNoteAndSendDescription(event: Electron.event, action: CreateStickerAction) {

        if (this.notes.has(action.id) && this.notes.get(action.id)) {
            let win: BrowserWindow = this.notes.get(action.id);
            win.focus();
            return;
        }

        let win: BrowserWindow = new BrowserWindow(this._windowOptions);

        win.loadURL(`file://${__dirname}/note.html`);

        if (process.env.NODE_ENV === 'development') {
            win.webContents.openDevTools();
        }

        win.webContents.on('dom-ready', () => {
            win.webContents.send(IpcChannels.editSticker, action.description);
        });
        this.notes.set(action.id, win);
        win.on('closed', () => this._removeTodoNotes(action.id));
    }

    _updateTodoNoteDescription(event: Electron.event, action: EditStickerAction) {
        let id = action.id;
        let win: BrowserWindow = this.notes.get(id);
        if (!win) {
            this.notes.delete(id);
            return;
        }
        win.webContents.send(action.channel, action.description);
    }

    _closeTodoNote(event: Electron.event, action: DeleteStickerAction) {
        let win: BrowserWindow = this.notes.get(action.id);
        if (!win) {
            this.notes.delete(action.id);
            return;
        }
        
        win.destroy();
    }

    _removeTodoNotes(id: string) {
        if (!this.notes.has(id)) {
            return;
        }
        this.notes.delete(id);
    }
}

const stickerHandler = new StickerHandler();
export default stickerHandler;