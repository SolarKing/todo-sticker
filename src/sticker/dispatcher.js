// @flow
import { type StickerActions } from './action';
import { ipcRenderer } from 'electron';

class Dispatcher {
    dispatch(action: StickerActions) {
        ipcRenderer.send(action.channel, action);
    }
}
const stickerDispatcher = new Dispatcher();

export default stickerDispatcher;