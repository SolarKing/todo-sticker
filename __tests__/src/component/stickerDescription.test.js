// @flow
import React from 'react';
import { shallow, type ShallowWrapper } from 'enzyme';
import StickerDescription from '../../../src/component/stickerDescription';
import { IpcChannels } from '../../../src/sticker/channel';
import { ipcRenderer } from 'electron';

jest.mock('electron', () => {
    let _listener;
    let ipcRenderer = {
        on: (channel: string, listener: Function) => {
            _listener = listener;
        },
        send: (channel: string, message: string) => {
            _listener({}, message);
        }
    };
    return {
        ipcRenderer,
    };
});

describe('StickerDescription', () => {
    let wrapper: ShallowWrapper;

    beforeAll(() => {
        wrapper = shallow(
            <StickerDescription/>
        );
    });

    it('should render correct', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should set description when ipcRenderer send new message', () => {
        ipcRenderer.send(IpcChannels.editSticker, '123');
        expect(wrapper.state().description).toEqual('123');
    });

});
