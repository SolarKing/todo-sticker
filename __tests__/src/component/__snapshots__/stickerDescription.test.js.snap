// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StickerDescription should render correct 1`] = `<stickerDescription__StickerDescriptionDiv />`;
